### ISSUE TITLE
How do I start using Ontopia REST?

### ISSUE THREAD

**Issue by boticello** [ontopia/ontopia#540]
I'd like to experiment with using Ontopia as a REST-based backend for my knowledgebase. How do I get started? Is it in the main Ontopia distribution? Is there any documentation?

**Comment by qsiebers** [comment]
Hey Boticello,

Good to hear that you want to start experimenting with the REST backend. 

> Is it in the main Ontopia distribution?

Not in a released version yet. It is part of the main ontopia repository though, if you are familiar with building maven projects from source then you can [start using it that way](https://github.com/ontopia/ontopia/wiki/HowToBuild). You can also use the [nightly snapshot builds](https://oss.sonatype.org/content/repositories/snapshots/net/ontopia/ontopia-rest/5.3.1-SNAPSHOT/) as a dependency in your project. If you are not comfortable with these options, I might be able to produce you a nightly build distribution zip.

> Is there any documentation?

Uhmm, nope. This is one of [the big todo's](https://ontopia.wordpress.com/2017/06/19/introducing-ontopia-rest/) on the list.

I'll leave this open as a reminder that we need documentation, but I suggest you add yourself to the [ontopia mailing list](https://groups.google.com/forum/#!forum/ontopia) and don't hesitate to ask for more help there if you get stuck. I'd love to hear about your experiences with ontopia REST.

**Comment by boticello** [comment]
Thank you Quintin. I'm not really comfortable with either option. I've run Ontopia in the past through Tomcat, which has been fairly straightforward – download and run. I'm not a Java guy.

I was on the Ontopia mailing list but have deleted my google accounts, so I will have to follow Ontopia news sporadically.

Perhaps it is better that I wait for a release version with some documentation. My aim is to use it as a general backend with a Vue.js frontend. I find the Topic Maps metamodel to be the most intuitive and powerful I've encountered, and I cannot understand why it is so obscure and forgotten. But I am not a daily developer and will be discouraged if I can't make sense of it. That's been my frustration – there are so many easy-to-use and well-documented backends in the javascript ecosystem, but none of them have the power of Topic Maps.

**Comment by achea** [comment]
Greetings,

In lieu of official documentation, do you have any tips to get started using this module?  What I'm hoping to accomplish is to set up 1 or more REST APIs backed by a RDBMS (like PostgreSQL) possibly with an initial topic map(s), and afterwards I can inspect the test cases for endpoints.  Like the fellow before, I'm not experienced with Tomcat but did find a web.xml in the ontopia-rest test directory.  All the ontopia-rest tests pass with `mvn test` so it's likely it's being used, I'm just not sure how to invoke my own.  I also don't see ontopia-rest in the ontopia-distribution-tomcat/libs/ folder that's generated, is it not bundled by default yet?

Any advice is appreciated, thanks!

P.S. I am still having trouble posting to the mailing list; I am fine with moving this discussion to the mailing list, just can't post there yet.

**Comment by qsiebers** [comment]
> I also don't see ontopia-rest in the ontopia-distribution-tomcat/libs/ folder that's generated, is it not bundled by default yet?

Good point, I'll see to it that it is included in the distribution. It might even be interesting to get a running REST API into the distribution for people to play with.

> In lieu of official documentation, do you have any tips to get started using this module? [...] but did find a web.xml in the ontopia-rest test directory

The tests are run on a very lean webapplication via a temporary Jetty server instance. You can have a look at [this application](https://github.com/ontopia/ontopia/tree/master/ontopia-rest/src/test/webapp) to see how to start with REST. If you'd deploy this application on the `test` context then you should be able to access the REST api at `http://localhost:8080/test/api`. For all the resources, methods and parameters you can use, the test cases are currently the best place to get information. Just a quick example: the [`TopicResourceGETTest`](https://github.com/ontopia/ontopia/blob/master/ontopia-rest/src/test/java/net/ontopia/topicmaps/rest/v1/topic/TopicResourceGETTest.java) class issues, amongst others, a request:
```
GET application/json http://localhost:8080/test/api/v1/topics/1
```

> P.S. I am still having trouble posting to the mailing list; I am fine with moving this discussion to the mailing list, just can't post there yet.

The mailing list is at http://groups.google.com/group/ontopia. Once joined (requires google account) you can post a new message there, or directly to ontopia@googlegroups.com.

**Comment by achea** [comment]
Thanks!  After a bit of tinkering, I've started an instance of the servlet in a not-ideal way.

```
diff --git a/ontopia-rest/pom.xml b/ontopia-rest/pom.xml
index 9475bcbc..189a902d 100644
--- a/ontopia-rest/pom.xml
+++ b/ontopia-rest/pom.xml
@@ -87,9 +87,17 @@
                                <configuration>
                                        <stopPort>9999</stopPort>
                                        <stopKey>ontopia</stopKey>
+                                                       <httpConnector>
+                                                               <port>${ontopia.jetty.port}</port>
+                                                       </httpConnector>
+                                                       <webAppSourceDirectory>${basedir}/src/test/webapp</webAppSourceDirectory>
+                                                       <useTestClasspath>true</useTestClasspath>
+                                                       <webApp>
+                                                               <contextPath>/test</contextPath>
+                                                       </webApp>
                                </configuration>
                                <executions>
-                                       <execution>
+                                       <!--execution>
                                                <id>start-jetty</id>
                                                <phase>test-compile</phase>
                                                <goals>
@@ -105,7 +113,7 @@
                                                                <contextPath>/</contextPath>
                                                        </webApp>
                                                </configuration>
-                                       </execution>
+                                       </execution-->
                                        <execution>
                                                <id>stop-jetty</id>
                                                <phase>test</phase>
```
What this does is configures the jetty plugin with all the settings in the execute start-jetty block, and then *disables* that start-jetty block as it installs a conflicting shutdown hook.  Then I can execute `mvn jetty:run` to start the test server.

Navigating to the root API URL prints a nice overview of the endpoints as well:
![Ontopia REST API endpoints](https://i.imgur.com/uxAvrtC.png)

Obviously, disabling the test-compile/start-jetty hook is bad, but it temporarily works.  I'm inexperienced with these Java Servlets though so I'll see how else to run this without jetty or maven.

A good step! :thumbsup:

Re mailing list - I apparently have to wait for my posts to be reviewed, even after joining in Sept. 2017.  Do you happen to have moderator privileges there to approve me?

**Comment by qsiebers** [comment]
> Re mailing list - I apparently have to wait for my posts to be reviewed, even after joining in Sept. 2017. Do you happen to have moderator privileges there to approve me?

I was unaware of this, I'm not sure who owns it but I'll try to hunt that info down. Your messages just came thru.
